src/net/client/cache.rs

```diff
@@ -0,0 +1,1177 @@
+where
+    CR: Send + Sync,
+    Upstream: Send + Sync,
+    C: Clock + Send + Sync,
+{
+    /// State of the request.
+    state: RequestState,
+
+    /// The request message.
+    request_msg: CR,
+
+    /// The upstream transport of the connection.
+    upstream: Upstream,
+
+    /// The cache of the connection.
+    cache: Cache<Key, Arc<Value<C>>>,
+
+    /// The configuration of the connection.
+    config: Config,
+
+    /// The clock to use for expiring cache entries.
+    clock: C,
+}
+
+impl<CR, Upstream, C> Request<CR, Upstream, C>
+where
+    CR: Clone + ComposeRequest + Send + Sync,
+    Upstream: SendRequest<CR> + Send + Sync,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    /// Create a new Request object.
+    fn new(
+        request_msg: CR,
+        upstream: Upstream,
+        cache: Cache<Key, Arc<Value<C>>>,
+        config: Config,
+        clock: C,
+    ) -> Request<CR, Upstream, C> {
+        Self {
+            state: RequestState::Init,
+            request_msg,
+            upstream,
+            cache,
+            config,
+            clock,
+        }
+    }
+
+    /// This is the implementation of the get_response method.
+    ///
+    /// This function is cancel safe.
+    async fn get_response_impl(&mut self) -> Result<Message<Bytes>, Error> {
+        loop {
+            match &mut self.state {
+                RequestState::Init => {
+                    let msg = self.request_msg.to_message()?;
+                    let header = msg.header();
+                    let opcode = header.opcode();
+
+                    // Extract Qname, Qclass, Qtype
+                    let mut question_section = msg.question();
+                    let question = match question_section.next() {
+                        None => {
+                            // No question. Just forward the request.
+                            let request = self
+                                .upstream
+                                .send_request(self.request_msg.clone());
+                            self.state =
+                                RequestState::GetResponseNoCache(request);
+                            continue;
+                        }
+                        Some(question) => question?,
+                    };
+                    if question_section.next().is_some() {
+                        // More than one question. Just forward the request.
+                        let request = self
+                            .upstream
+                            .send_request(self.request_msg.clone());
+                        self.state =
+                            RequestState::GetResponseNoCache(request);
+                        continue;
+                    }
+                    let qname = question.qname();
+                    let qclass = question.qclass();
+                    let qtype = question.qtype();
+
+                    if !(opcode == Opcode::Query && qclass == Class::In) {
```

Comment:
I'm confused, which test?

Reply:
I’m now, too. You could quickly look at the opcode and forward if that is not QUERY even before looking at the question and only then get the question and look at that.

Reply:
I don't see a nice way of doing that. Just checking for opcode early would duplicate the send_request call. It would be nice to skip extracting qname/qclass/qtype if opcode is not Query, but it seems that it would make that code more complex. In addition, anything that is not using opcode == Query would probably not use the cache in the first place.